core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala

```diff
@@ -0,0 +1,485 @@
+    val clientAddr = RpcAddress(addr.getHostName, addr.getPort)
+    val broadcastMessage =
+      synchronized {
+        // If the first connection to a remote RpcEnv is found, we should broadcast "Associated"
+        if (remoteAddresses.put(clientAddr, remoteEnvAddress).isEmpty) {
+          // clientAddr connects at the first time
+          val count = remoteConnectionCount.getOrElse(remoteEnvAddress, 0)
+          // Increase the connection number of remoteEnvAddress
+          remoteConnectionCount.put(remoteEnvAddress, count + 1)
+          if (count == 0) {
+            // This is the first connection, so fire "Associated"
+            Some(Associated(remoteEnvAddress))
+          } else {
+            None
+          }
+        } else {
+          None
+        }
+      }
+    broadcastMessage.foreach(dispatcher.broadcastMessage)
+    dispatcher.postMessage(requestMessage, callback)
+  }
+
+  override def getStreamManager: StreamManager = new OneForOneStreamManager
+
+  override def exceptionCaught(cause: Throwable, client: TransportClient): Unit = {
+    val addr = client.getChannel().remoteAddress().asInstanceOf[InetSocketAddress]
+    if (addr != null) {
+      val clientAddr = RpcAddress(addr.getHostName, addr.getPort)
+      val broadcastMessage =
+        synchronized {
+          remoteAddresses.get(clientAddr).map(AssociationError(cause, _))
+        }
+      if (broadcastMessage.isEmpty) {
+        logError(cause.getMessage, cause)
+      } else {
+        dispatcher.broadcastMessage(broadcastMessage.get)
+      }
+    } else {
+      // If the channel is closed before connecting, its remoteAddress will be null.
+      // See java.net.Socket.getRemoteSocketAddress
+      // Because we cannot get a RpcAddress, just log it
+      logError("Exception before connecting to the client", cause)
+    }
+  }
+
+  override def connectionTerminated(client: TransportClient): Unit = {
+    val addr = client.getChannel().remoteAddress().asInstanceOf[InetSocketAddress]
+    if (addr != null) {
+      val clientAddr = RpcAddress(addr.getHostName, addr.getPort)
+      val broadcastMessage =
+        synchronized {
+          // If the last connection to a remote RpcEnv is terminated, we should broadcast
+          // "Disassociated"
+          remoteAddresses.get(clientAddr).flatMap { remoteEnvAddress =>
+            remoteAddresses -= clientAddr
+            val count = remoteConnectionCount.getOrElse(remoteEnvAddress, 0)
+            assert(count != 0, "remoteAddresses and remoteConnectionCount are not consistent")
+            if (count - 1 == 0) {
+              // We lost all clients, so clean up and fire "Disassociated"
+              remoteConnectionCount.remove(remoteEnvAddress)
+              Some(Disassociated(remoteEnvAddress))
+            } else {
+              // Decrease the connection number of remoteEnvAddress
+              remoteConnectionCount.put(remoteEnvAddress, count - 1)
```

Comment:
> There might be an edge case here, in case connections are lost without an explicit clientFactory.close() call; in those cases it might be possible that you ping-pong back and forth between 0 and 1, thus sending multiple Associated and Disassociated messages.

Why connection lost will make its counter change from 0 to 1? Could you clarify it?